### PR TITLE
Add demographics validation

### PR DESCRIPTION
**Pull request by imranmaj**
https://sagebionetworks.jira.com/browse/BRIDGE-3030 and https://sagebionetworks.jira.com/browse/BRIDGE-3352

### CODE REVIEW

File: src/test/java/org/sagebionetworks/bridge/json/DemographicUserSerializerDeserializerTest.java [hunk 1]
```diff
@@ -176,9 +197,9 @@ public void deserializeNullDemographicValue() throws JsonMappingException, JsonP
     public void setNullDemographicValueInner() {
         Demographic demographic = new Demographic();
         List<DemographicValue> values = new ArrayList<>();
-        values.add(new DemographicValue(null));
+        values.add(new DemographicValue((String) null));
```
Comment (imranmaj) [inline src/test/java/org/sagebionetworks/bridge/json/DemographicUserSerializerDeserializerTest.java]: Remove

File: src/test/java/org/sagebionetworks/bridge/json/DemographicValueDeserializerTest.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        test("0.1234567890123456", "0.1234567890123456");
+    }
+
+    @Test
+    public void decimalNegFractionGreaterThanNegOne() throws JsonMappingException, JsonProcessingException {
+        test("-0.1234567890123456", "-0.1234567890123456");
+    }
+
+    @Test
+    public void decimalPosLarge() throws JsonMappingException, JsonProcessingException {
```
Comment (imranmaj) [inline src/test/java/org/sagebionetworks/bridge/json/DemographicValueDeserializerTest.java]: Wrong name

File: src/test/java/org/sagebionetworks/bridge/services/DemographicServiceTest.java [hunk 3]
```diff
@@ -91,6 +115,286 @@ public void saveDemographicUser() {
...
+        Demographic demographic = new Demographic(TEST_APP_ID, demographicUser, "bar", false,
+                ImmutableList.of(new DemographicValue("random value")), null);
+        demographicUser.setDemographics(ImmutableMap.of("bar", demographic));
+
+        // execute
+        demographicService.saveDemographicUser(demographicUser, account);
+
+        // verify
+        verify(appConfigElementService).getMostRecentElements(TEST_APP_ID, false);
+        // nothing else should happen because the app config element key does nkot match
```
Comment (imranmaj) [inline src/test/java/org/sagebionetworks/bridge/services/DemographicServiceTest.java]: Typo

File: src/test/java/org/sagebionetworks/bridge/models/studies/DemographicValuesValidationConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.exc.MismatchedInputException;
+
+public class DemographicValuesValidationConfigurationTest {
+    @Test
+    public void serialize() throws JsonProcessingException {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/models/studies/DemographicValuesValidationConfigurationTest.java]: Do we have a test where we deserialize a valid validation config?

File: src/test/java/org/sagebionetworks/bridge/services/DemographicServiceTest.java [hunk 5]
```diff
@@ -91,6 +115,286 @@ public void saveDemographicUser() {
...
+        DemographicUser demographicUser = new DemographicUser("test-id", TEST_APP_ID, null, TEST_USER_ID, null);
+        Demographic demographic = new Demographic(TEST_APP_ID, demographicUser, "category", false,
+                ImmutableList.of(new DemographicValue("foo")), null);
+        demographicUser.setDemographics(ImmutableMap.of("category", demographic));
+
+        // execute
+        demographicService.saveDemographicUser(demographicUser, account);
+
+        // verify
+        verify(appConfigElementService).getMostRecentElements(TEST_APP_ID, false);
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/services/DemographicServiceTest.java]: We should also verify that the saved demographics have no invalidity.

File: src/main/java/org/sagebionetworks/bridge/validators/DemographicValuesValidator.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ * This does not use the Spring validator framework because errors are stored in
+ * the validated object (Demographic) directly instead of being returned.
+ */
+public class DemographicValuesValidator {
+    private static final String DEMOGRAPHICS_ENUM_DEFAULT_LANGUAGE = "en";
+    private static final String INVALID_DATA = "invalid data";
+    // currently not exposing configuration errors
+    public static final String INVALID_CONFIGURATION = INVALID_DATA;
+    private static final String INVALID_CONFIGURATION_BAD_LANGUAGE_CODE = INVALID_CONFIGURATION;
+    private static final String INVALID_CONFIGURATION_MIN_LARGER_THAN_MAX = INVALID_CONFIGURATION;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/validators/DemographicValuesValidator.java]: Mentioned in code review that we shouldn't hide invalid configuration. If we're moving this to its own API, we can use the same Validator pattern we use for other APIs and ensure that our configuration as always valid.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
In addition, changes to move DemographicsValidationConfig to its own storage and APIs, and enabling validation for all demographics, app-level and study-level.

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java [hunk 7]
```diff
@@ -339,4 +352,73 @@ public PagedResourceList<DemographicUser> getDemographicUsers(
...
+            "/v3/participants/demographics/validation/{categoryName}" })
+    public DemographicValuesValidationConfig saveValidationConfig(
+            @PathVariable(required = false) Optional<String> studyId, @PathVariable String categoryName)
+            throws BadRequestException, NotAuthenticatedException, UnauthorizedException {
+        String studyIdNull = studyId.orElse(null);
+
+        UserSession session;
+        if (studyId.isPresent()) {
+            // study level demographics
+            session = getAuthenticatedSession(Roles.RESEARCHER, Roles.STUDY_COORDINATOR);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java]: Should add `CAN_UPDATE_STUDIES.checkAndThrow(STUDY_ID, studyId);` See StudyController.updateStudy() for an example.

Also, this should be DEVELOPER and STUDY_DESIGNER, not RESEARCHER and STUDY_COORDINATOR.

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java [hunk 8]
```diff
@@ -339,4 +352,73 @@ public PagedResourceList<DemographicUser> getDemographicUsers(
...
+            "/v3/participants/demographics/validation/{categoryName}" })
+    public DemographicValuesValidationConfig getValidationConfig(
+            @PathVariable(required = false) Optional<String> studyId, @PathVariable String categoryName)
+            throws EntityNotFoundException, NotAuthenticatedException, UnauthorizedException {
+        String studyIdNull = studyId.orElse(null);
+
+        UserSession session;
+        if (studyId.isPresent()) {
+            // study level demographics
+            session = getAuthenticatedSession(Roles.RESEARCHER, Roles.STUDY_COORDINATOR);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java]: Similar, this should use `CAN_READ_STUDIES.checkAndThrow(STUDY_ID, studyId);` See StudyController.getStudy() for example.

We can make this available to RESEARCHER, DEVELOPER, STUDY_COORDINATOR, and STUDY_DESIGNER.

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java [hunk 9]
```diff
@@ -339,4 +352,73 @@ public PagedResourceList<DemographicUser> getDemographicUsers(
...
+            "/v3/participants/demographics/validation/{categoryName}" })
+    public StatusMessage deleteValidationConfig(
+            @PathVariable(required = false) Optional<String> studyId, @PathVariable String categoryName)
+            throws EntityNotFoundException, NotAuthenticatedException, UnauthorizedException {
+        String studyIdNull = studyId.orElse(null);
+
+        UserSession session;
+        if (studyId.isPresent()) {
+            // study level demographics
+            session = getAuthenticatedSession(Roles.RESEARCHER, Roles.STUDY_COORDINATOR);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java]: This should have the same CAN_UPDATE_STUDIES check and permissions as saveValidationConfig().

File: src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java [hunk 10]
```diff
@@ -339,4 +352,73 @@ public PagedResourceList<DemographicUser> getDemographicUsers(
...
+            throws BadRequestException, NotAuthenticatedException, UnauthorizedException {
+        String studyIdNull = studyId.orElse(null);
+
+        UserSession session;
+        if (studyId.isPresent()) {
+            // study level demographics
+            session = getAuthenticatedSession(Roles.RESEARCHER, Roles.STUDY_COORDINATOR);
+        } else {
+            // app level demographics
+            session = getAuthenticatedSession(Roles.ADMIN);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/spring/controllers/DemographicController.java]: App-wide permissions should be DEVELOPER. This applies to all 3 of the APIs.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Looks good. Just comments about permissions.

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**